### PR TITLE
[[ Bug 14435 ]] Allow setting a filename path that is relative to the de...

### PR DESCRIPTION
**Pull request by livecodepanos**
...faultFolder

### CODE REVIEW

File: engine/src/player-platform.cpp [hunk 1]
```diff
@@ -1437,7 +1449,15 @@ Exec_stat MCPlayer::setprop(uint4 parid, Properties p, MCExecPoint &ep, Boolean
 	{
 #ifdef /* MCPlayer::setprop */ LEGACY_EXEC
         case P_FILE_NAME:
-            if (filename == NULL || data != filename)
+        {
+            // Edge case: Suppose filenameA is a valid relative path to defaultFolderA, but invalid relative path to defaultFolderB
+            // 1. Set defaultFolder to defaultFolderB. Set the filename to filenameA. Video will become empty, since the relative path is invalid.
+            // 2. Change the defaultFolder to defaultFolderA. Set the filename again to filenameA. Now the relative path is valid
+            char *t_resolved_filename;
+            resolveplayerfilename(filename, t_resolved_filename);
```
Comment (runrevmark) [inline engine/src/player-platform.cpp]: YOu aren't checking the return value here... t_resolved_filename could be undefined after this call if it returns false.

File: engine/src/player-platform.cpp [hunk 2]
```diff
@@ -1437,7 +1449,19 @@ Exec_stat MCPlayer::setprop(uint4 parid, Properties p, MCExecPoint &ep, Boolean
...
 #ifdef /* MCPlayer::setprop */ LEGACY_EXEC
         case P_FILE_NAME:
-            if (filename == NULL || data != filename)
+        {
+            // Edge case: Suppose filenameA is a valid relative path to defaultFolderA, but invalid relative path to defaultFolderB
+            // 1. Set defaultFolder to defaultFolderB. Set the filename to filenameA. Video will become empty, since the relative path is invalid.
+            // 2. Change the defaultFolder to defaultFolderA. Set the filename again to filenameA. Now the relative path is valid
+            char *t_resolved_filename;
+            bool t_success = false;
+            t_success = resolveplayerfilename(filename, t_resolved_filename);
```
Comment (runrevmark) [inline engine/src/player-platform.cpp]: I think this needs to be done slightly differently.

The player probably needs to store the resolved (absolute) filename to compare with.

It should resolve here and if the resolved filename is different from the one currently being used, it needs to re-prepare.

You mustn't re-resolve the path on prepare though as things might have changed in between - you need to only resolve once and use that filename. Thus, you'll need to make a variant of prepare that takes the filename to use.

The filename property should still reflect what it was set to, but prepare should be responsible for setting the internal 'absolute' path.

File: engine/src/player-platform.cpp [hunk 3]
```diff
@@ -1823,6 +1855,8 @@ IO_stat MCPlayer::save(IO_handle stream, uint4 p_part, bool p_force_ext)
 			return stat;
 		if ((stat = IO_write_string(filename, stream)) != IO_NORMAL)
 			return stat;
+        if ((stat = IO_write_string(resolved_filename, stream)) != IO_NORMAL)
```
Comment (runrevmark) [inline engine/src/player-platform.cpp]: You can't do this as it changes the file format!

The player object should save 'filename' as what was set as the property; but use the 'resolved' filename at the point the movie is prepared.

File: engine/src/player-platform.cpp [hunk 4]
```diff
@@ -881,6 +882,7 @@ MCPlayer::MCPlayer(const MCPlayer &sref) : MCControl(sref)
 {
 	nextplayer = NULL;
 	filename = strclone(sref.filename);
+    resolved_filename = strclone(sref.resolved_filename);
```
Comment (runrevmark) [inline engine/src/player-platform.cpp]: Don't clone the resolved_filename - it should be reinitialized at the appropriate time... A cloned object won't necessarily be on the same stack and so the resolved_filename might be different.

**Comment by runrevmark** [comment]
I think it is worth checking the behavior of < 6.7 here. Pre-6.7 when did the 'filename' property get resolved to a full path? The current code here means it only gets computed on setting the fileName - however, I think it should happen when it is needed (i.e. when it triggers a movie to load, maybe OnOpen).

**Comment by livecodepanos** [comment]
Indeed, in pre-6.7 the 'filename' property got resolved to a full path in prepare(). I just updated this pull request.

**Comment by runrevmark** [comment]
Reviewed-by: @runrevmark.

**Comment by peter-b** [comment]
@livecode-vulcan review by @runrevmark okay 37ed7e5

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok 37ed7e58eb910df553ca1de7320747eed83e2e5a

**Pull request merged**